phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java

```diff
@@ -216,22 +234,146 @@ public void close() throws SQLException {
                 }
             }
         }
-        
+    }
+    
+    public ServerCache checkServerCache(final byte[] cacheId, ScanRanges keyRanges, final TableRef cacheUsingTableRef,
```

Comment:
I am thinking here, can we do this differently?
Instead of making RPC calls of "checkServerCache" for the first and every subsequent queries, we do NOT make any calls, neither "check" or "add" when the persistent-cache hint is available and catches the {{PersistentCacheNotFoundException}} on the first attempt (or later attempts if somehow the cache has been evicted) and then try adding the cache all over again. I think it will be more efficient in general.

Reply:
FYI, we already have a mechanism for this. If the server throws a HashJoinCacheNotFoundException, then the client will react by sending the hash cache to the region servers that don't have it.

Reply:
Yea makes sense. For this, I'm thinking of modifying the retry logic in BaseResultIterators to disable persistent caching on retries. I'm thinking I can add a get/set'er pair to the "context" that can be used for this